proposals/factory_constructors.md

```diff
@@ -0,0 +1,216 @@
+
+```cs
+public class Widget
+{
+    public string Name { get; set; }
+    public int Id { get; init; }
+
+    internal Widget() { }
+}
+
+public static class WidgetFactory
+{
+    public static factory Widget() => new Widget();
+}
+
+var w = new WidgetFactory.Widget()
+{
+    Name = "333fred",
+    Id = 43
+};
+```
+
+### Grammar
+
+Factory constructors are specified by the following grammar:
+
+```antlr
+factory_constructor_declaration
+    : attributes? factory_constructor_modifier* 'factory' factory_constructor_declarator factory_constructor_body
+    ;
+
+factory_constructor_modifier
+    : 'new'
+    | 'public'
+    | 'protected'
+    | 'internal'
+    | 'private'
+    | 'static'
+    | 'virtual'
+    | 'sealed'
+    | 'override'
+    | 'abstract'
+    | factory_constructor_modifier_unsafe
+    ;
+
+factory_constructor_modifier_unsafe
+    : 'unsafe'
+    ;
+
+factory_constructor_declarator
+    : type '(' formal_parameter_list? ')'
+    ;
+
+factory_constructor_body
+    : block
+    | ';'
+    ;
+```
+
+Some examples:
+
+```cs
+public static interface I
+{
+    int MyProperty { get; init; }
+    // Define a copy constructor for the ability to `with` any `I` implementation
+    factory I();
+}
+public abstract class AbsI : I
+{
+    public int MyProperty { get; init; }
+
+    protected abstract new factory I();
+    factory I.I()
+        => new this.I() // `this.I` is required to disabmiguate from the default parameterless constructor
+           {
+                MyField = this.MyField
+           };
+}
+public abstract class DerivedI : AbsI
+{
+    public int DerivedProperty { get; set; }
+    protected override factory I() => new DerivedProperty() { DerivedProperty = this.DerivedProperty };
+}
+
+public class IFactory
+{
+    public factory I(int param1, int param2) => ...;
+    public factory I(string param1) => ...;
+    public static factory I(object param1) => ...;
+}
+```
+
+Factory constructors are invoked by calling `new`, and providing a primary expression that resolves to a factory constructor.
+
+```antlr
+object_creation_expression
+    : 'new' primary_expression_or_type '(' argument_list? ')' object_or_collection_initializer?
+    | 'new' primary_expression_or_type object_or_collection_initializer
+    ;
+
+primary_expression_or_type
+    : primary_expression
+    | type
+    ;
+```
+
+For an object creation expression, the compiler will first attempt to bind the `primary_expression_or_type` as a type. If it binds,
+and if overload resolution with the provided arguments returns anything other than an empty set, then no attempt is made to bind
+the `primary_expression_or_type` as a `primary_expression`. However, if it fails to bind as a type or overload resolution fails to
+find an applicable constructor, the compiler will bind the `primary_expression_or_type` as a `primary_expression`, and attempt to
```

Comment:
i would remove the part about overload resolution.  if it's a type, period, treat it as a type.  It would be *very* poor form to have ambiguity here IMO where the types/factories were name the same.

Reply:
I don't think that's true, actually, and it's kind of central to the proposal. A type can define static factory constructors that return a new instance of itself. It's definitely a weakness to the proposal that needs to be hammered on some more, but I don't think we can straight-up throw it out without removing a large portion of the gain this proposal brings.

Reply:
can you give an actual example of this?  i.e. both the production side and hte consumption side?  i think there are issues here, but havign a concrete example woudl help hammer them out.  Thanks!

Reply:
Well, for example, something like this:

```cs
class Widget
{
    static factory Widget() => new Widget(string.Empty, -1);
    static factory Widget(Widget widget) => new Widget() { Title = widget.Title, Id = widget.Id };
    private Widget(string title, int id) => (Title, Id) = (title, id);
    public string Title { get; init; }
    public int Id { get; set; }
}
```

Maybe a bit contrived, but I think the idea is simple enough: One static factory wants to chain to another in the same class. We don't require you to specify the containing class for calling methods in the same class today, so it would be strange for us to require this for constructor calls.